forem/forem
# Follow and Hidden Buttons on /tags page

## What type of PR is this? (check all applicable)


## Description
On the Tags page (/tags), each Tag will show an additional "Hide" toggle, which will "follow" the tag for the user and set the tag weight to -1 on the tag. This will show for both "Top Tags" and the tag search results.

- [x]  On the Tags page, a tag can either be:
  - [x]  Default, unfollowed and unhidden, in which both a "Follow" and "Hide" button will show
  - [x]  Followed, in which a "Following" toggle button will show, which, when pressed, will unfollow the tag.
  - [x]  Hidden, in which an "Unhide" toggle button will show, which, when pressed, will unhide the tag.


**Some technical context:** 

Previously we were using the [`followButtons.js`](https://github.com/forem/forem/blob/b423f92659d569eb69951610b67ea2b0d8ffc611/app/javascript/packs/followButtons.js#L1) to toggle the state of the following of tags on the /tags page. In fact, `followButtons.js` is essentially a JavaScript component that is used to manage **all** the followButtons across the application. Whilst this may have worked in the past, we've now introduced a "Hide" button which has affects on the "follow" button, hence causing the the /tags page follow button now deviate form this standard pattern. Trying to retrofit this pattern into the existing `followButtons.js` does not seem feasible. You can find my [very draft, very ugly attempt](https://github.com/forem/forem/pull/19921).

In addition, `followButtons.js` does a lot of manual DOM manipulation for changing the state of the buttons based on the current state. This seems like a better use case to be handled by Preact rather than doing the DOm manipulation by hand in vanilla JS. Hence, you'll notice that I created a Preact component for the buttons.

_In the future: _
It's a bit weird to be in this intermediate state where most of the Follow buttons are being handled by `followButtons.js` but then the /tags page has its own mechanism to handle these buttons. I'd like for us to improve on followButtons and turn most of that manual code into a Preact component that will manage the state of the DOM more easily and make it more readable and easy to follow. I'll do a further write-up and chat more about it in our upcoming snack n study with the team. 

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/19675
- Closes #

## QA Instructions, Screenshots, Recordings

- Go to /tags and click the "follow" button to follow tags. Refresh the page and ensure that they stay followed. 
- Now click on the "Following" button to unfollow the tags.  Refresh the page and ensure that they stay unfollowed.
- Now find a tag card with a "Follow" button and a "Hide" button. Click on the "Hide" button to to hide the tag. The "Follow" Button should disappear and an "Unhide" button should show up in place of the "Hide" button.
- Now find a tag card with a "Following" button and a "Hide" button. Click on the "Hide" button to to hide the tag. The "Following" Button should disappear and an "Unhide" button should show up in place of the "Hide" button.
- Now click on the "Unhide" Button to unhide the tags. The "Following" (not the "Follow") button appear and the "Hide" button will appear.

https://www.loom.com/share/b2c516e897674c2193caf4554bb70b6a?sid=f8f104ef-a86c-4e73-a5bc-02ca803f8a3d

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/UpJQc5hFqjp23nviMG/giphy.gif)
